### ISSUE TITLE
Transit Secrets Engine BatchedEncryption key_version and nonce are nil

### ISSUE THREAD

**Issue by bentdan** [rajanadar/VaultSharp#184]
**Describe the bug**
Getting VaultSharp.Core.VaultApiException with errors saying key_version and nonce are unconvertible type'\u003cnil\u003e' when not set and trying to use defaults (using code similar to what you have in sample and in readme).  If Nonce and KeyVersion are manually set in EncryptionItem then I only get the error for KeyVersion (as shown in code snippet below).

**VaultSharp Version**
1.6.2.1

**Vault Version**
1.5.5 Enterprise

**Does this work with Vault CLI?**
Have not tried

**Sample Code Snippet**
```cs
public async Task<IEnumerable<string>> EncryptBatchAsync(Dictionary<string, string> plainTextWithDerivationContext)
{
            var nonce = Convert.ToBase64String(Enumerable.Range(0, 12).Select(i => (byte)i).ToArray());
            var batchedEncryptionItems = new List<EncryptionItem>();
            foreach(KeyValuePair<string, string> entry in plainTextWithDerivationContext)
            {
                var encodedPlainText = Convert.ToBase64String(Encoding.UTF8.GetBytes(entry.Key));
                var encryptionItem = new EncryptionItem {Base64EncodedPlainText = encodedPlainText};
                encryptionItem.Nonce = nonce;
                encryptionItem.KeyVersion = 1;
                if (entry.Value != null)
                {
                    var encodedContext = Convert.ToBase64String(Encoding.UTF8.GetBytes(entry.Value));
                    encryptionItem.Base64EncodedContext = encodedContext;
                }
                batchedEncryptionItems.Add(encryptionItem);
            };
            var encryptOptions = new EncryptRequestOptions
            {
                BatchedEncryptionItems = batchedEncryptionItems
            };

            Secret<EncryptionResponse> encryptionResponse = await VaultClient.V1.Secrets.Transit.EncryptAsync(vaultKeyRingName, encryptOptions);

            return encryptionResponse.Data.BatchedResults.Select(r => r.CipherText);
}
```

**Exception Details/Stack Trace/Error Message**
... threw exception: 
VaultSharp.Core.VaultApiException: {"errors":["1 error occurred:\n\t* failed to parse batch input: 3 error(s) decoding:\n\n* '[0].key_version' expected type 'int', got unconvertible type 'json.Number'\n* '[1].key_version' expected type 'int', got unconvertible type 'json.Number'\n* '[2].key_version' expected type 'int', got unconvertible type 'json.Number'\n\n"]}

**Any additional info**
Nope

**Comment by rajanadar** [comment]
will have a look this weekend.

**Comment by rajanadar** [comment]
Hi @bentdan 

It looks like we have hit a legit Vault Api Bug. The Go Json library used by Vault is not able to understand the incoming Json Number as a GoLang int type, resulting in a casting error.

- As per their documentation here, https://www.vaultproject.io/api/secret/transit#encrypt-data when a batch-input is provided, then the information will be picked up from the batch input items. The root level (single encrypt) item attributes will be ignored.
- As per their source code, this is the exact error line you are encountering:

```go
// https://github.com/hashicorp/vault/blob/bab5d07f01390d47fbeb79edc4b570e3153263f0/builtin/logical/transit/path_encrypt.go#L198

errs.Errors = append(errs.Errors, fmt.Sprintf("'[%d].key_version' expected type 'int', got unconvertible type '%T'", i, item["key_version"]))


```

- The logic flow for them is that, when they receive the request payload, they look at batch items first. 
    - And if it is not empty (Line 231), they parse it first into a list of items. And for every item, we are encountering the error, because their code is not able to understand that the number (key_version: 1) we sent in Json as Go int type. The casting code is as follows, where v is the Json Number type:

```
 if casted, ok := v.(int)    // Line 195
```
- However, if the batch items is empty, i.e. if the request payload is about the client wanting to encrypt just one item, then they parse out the root level fields of the ```EncryptRequestOptions``` object. The thing to note is that they use a different mechanism than parsing batch inputs to construct the same key_version field.

```go

// https://github.com/hashicorp/vault/blob/bab5d07f01390d47fbeb79edc4b570e3153263f0/builtin/logical/transit/path_encrypt.go#L251

KeyVersion: d.Get("key_version").(int)

```

- Apparently they have a comment here: https://github.com/hashicorp/vault/blob/bab5d07f01390d47fbeb79edc4b570e3153263f0/builtin/logical/transit/path_encrypt.go#L127

> // decodeBatchRequestItems is a fast path alternative to mapstructure.Decode to decode []BatchRequestItem.
   // It aims to behave as closely possible to the original mapstructure.Decode and will return the same errors.
   // https://github.com/hashicorp/vault/pull/8775/files#r437709722

- The ```mapstructure.Decode``` seems to work well. But not the alternative they have.
- So for some reason, they parse the same fields differently when it is a single encrypt item, and a batch, and the batch parsing of key_version field is messed up, since it doesn't understand a simple incoming json integer. 
- From VaultSharp's perspective, we are doing nothing special. We send the same key_version field and value, in single or batch case, and the single item case works as expected.

-----------------

Alright, what are next steps here:

This looks like a vault bug, so there is no real fix on VaultSharp side to do. I'll try to follow up with Vault team to confirm if it is a bug and get a response. That's an action item on me.

For you, can you please try out 2 things:

1. Can you please try out a batch encrypt via Vault CLI? It should be sending the same JSON as VaultSharp to Vault, and it should be hitting the same bug. If CLI works, then please send me the JSON you used for CLI. If CLI gives you the same errors, then it is definitely a Vault bug.
2. To immediately unblock you, I have a very crude workaround, but it'll work only if you are encrypting using your latest key version, not a specific key version. Here it is.

Basically, in VaultSharp, we are intercepting the outgoing batch request and removing the key_version field altogether. Presence of that field (with null value or 0 or 1 or any value), seems to mess up Vault. That field needs to absent from Json altogether. So use this code snippet and let me know.

```

var encryptOptions = new EncryptRequestOptions
{
    BatchedEncryptionItems = batchedEncryptionItems
};

VaultClient.Settings.BeforeApiRequestAction += (HttpClient httpClient, HttpRequestMessage httpRequestMessage) =>
{
    if (httpRequestMessage.RequestUri.AbsoluteUri.Contains("/encrypt"))
    {
        var originalRequestJson = JsonConvert.SerializeObject(encryptOptions);
        var jsonPartToReplace = "\"key_version\":1,";

        var modifiedRequestJson = originalRequestJson.Replace(jsonPartToReplace, string.Empty);

        httpRequestMessage.Content = new StringContent(
            modifiedRequestJson,
            Encoding.UTF8);
    }
};

Secret<EncryptionResponse> encryptionResponse = await VaultClient.V1.Secrets.Transit.EncryptAsync(vaultKeyRingName, encryptOptions);

```

**Comment by rajanadar** [comment]
I tried an independent Postman call to Vault Api as well and get the same error. There is already a Vault bug with this exact issue: https://github.com/hashicorp/vault/issues/10232

Let's see how it progresses.

**Comment by rajanadar** [comment]
@bentdan 

If you are always encrypting using the latest key, please use my above workaround. It'll ensure that your batch encryption is performant by doing all encryption in 1 round trip.

If you are encrypting using specific key versions, then the above workaround won't work, till the Vault team fixes the issue. I would suggest that, in that case, you loop through your items and encrypt one item at a time. Little less performant, but unblocks you.

I'll close this issue for now. Let me know if any more VaultSharp help is needed.

**Comment by bentdan** [comment]
Using VaultClient.Settings.BeforeApiRequestAction to replace does work.
I set it up to replace "\"key_version\":null," and  "\"nonce\":null,"

**Comment by rajanadar** [comment]
Cool.

**Comment by rajanadar** [comment]
Just be careful in your prod environment because if any dev changes json converter settings to pretty json or anything like that, the string replacement might fail due to spaces, beautification. String replacement is always bug prone. 

So you may want to always have a unit test your json converter with that same encrypt request object and always assert that version got replaced.

**Comment by rajanadar** [comment]
@bentdan 

The Vault team has fixed this issue in Vault 1.7 onwards. https://github.com/hashicorp/vault/issues/10232
Feel free to get rid of the "string substitution logic" when you upgrade vault.
